src/bsoncxx/validate.hpp

```diff
@@ -0,0 +1,136 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <cstdint>
+
+#include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+
+class validator;
+
+///
+/// Validates a BSON document. This is a simplified overload that will
+/// only do the bare minimum validation of document structure, and does
+/// not provide any further information if the document is found to be invalid.
+///
+/// @param data
+///   A buffer containing a BSON document to validate.
+/// @param length
+///   The size of the buffer.
+///
+/// @returns
+///   An engaged optional containing a view if the document is valid, or
+///   an unengaged optional if the document is invalid.
+///
+BSONCXX_API stdx::optional<document::view> validate(const std::uint8_t* data,
+                                                    std::size_t length);
+
+///
+/// Validates a BSON document. This overload provides additional control over the
+/// precise validation that is performed, and will give the caller access to the
+/// offset at which the document was found to be invalid.
+///
+/// @param data
+///   A buffer containing a BSON document to validate.
+/// @param length
+///   The size of the buffer.
+/// @param validator
+///   A validator used to configure what checks are done. If validation fails, it
+///   will contain the offset at which the document was found to be invalid.
+///
+/// @returns
+///   An engaged optional containing a view if the document is valid, or
+///   an unengaged optional if the document is invalid.
+///
+BSONCXX_API stdx::optional<document::view> validate(const std::uint8_t* data,
+                                                    std::size_t length,
+                                                    validator& validator);
```

Comment:
Lets make the invalid offset an out parameter:

BSONCXX_API stdx::optional<document::view> validate(
    const std::uint8_t* data,
    std::size_t length,
    const validator& validator = validator(),
    std::size_t** invalid_offset = NULL)

And then we can remove the friendship, and the member state.